Adjust Sugar Cane to break when there is no water

bonbionseker opened the pull request:
## Introduction
This PR allows Sugar Cane to break when it cannot find water.

https://github.com/user-attachments/assets/f1dec797-1eaf-4179-a057-0be68490c837
## Tests
I tested this PR by doing the following (tick all that apply):
- [x] Playtesting using a Minecraft client (provide screenshots or a video)

Commit:
Adjust Sugar Cane to break when there is no water

Commit:
Update Sugarcane.php

dktapps commented on src/block/Sugarcane.php:
```diff
@@ -123,4 +133,15 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
+		 $down = $this->getSide(Facing::DOWN);
+		 foreach(Facing::HORIZONTAL as $side){
+			 $sideBlock = $down->getSide($side);
+			 if($sideBlock instanceof Water || $sideBlock instanceof FrostedIce){
```
Really not supposed to use instanceof for this

bonbionseker replied:
then how can i do it .-.

jasonw4331 replied:
check the id value

bonbionseker replied:
Okay.. I used BlockTypeIds, is it correct now? ._.

dktapps reviewed (commented):
Is this vanilla behaviour?

ghost commented:
> vanilla behaviour?

yes it is true vanilla behavior :D

Commit:
Update Sugarcane.php

I used Block Type Ids instead of instanceof

kostamax27 commented on src/block/Sugarcane.php:
```diff
@@ -98,6 +100,14 @@ public function ticksRandomly() : bool{
 		if(!$this->getSide(Facing::DOWN)->hasSameTypeId($this)){
+			if(!$this->isSupportedByWater()) {
+				$world = $this->position->getWorld();
+				$world->addParticle($this->position, new BlockBreakParticle($this));
+				$world->addSound($this->position, new BlockBreakSound($this));
+				$world->useBreakOn($this->position);
```
maybe it will be better

```php
$world->useBreakOn($this->position, createParticles: true);
```

bonbionseker replied:
yes this is better :D

Commit:
Update Sugarcane.php

yes this is better

kostamax27 commented on src/block/Sugarcane.php:
```diff
@@ -98,6 +98,12 @@ public function ticksRandomly() : bool{
 
 	public function onRandomTick() : void{
 		if(!$this->getSide(Facing::DOWN)->hasSameTypeId($this)){
+			if(!$this->isSupportedByWater()) {
```
This should happen not only during onRandomTick but also if something nearby is updated, for example, when a block is placed nearby.

https://github.com/user-attachments/assets/d6becf94-1c18-46d3-aae2-ecfc2c13fd4a

bonbionseker replied:
I added onNearbyBlockChange for this thanks 👍

Commit:
I added onNearbyBlockChange

Commit:
Code Style Fix

kostamax27 commented on src/block/Sugarcane.php:
```diff
@@ -123,4 +129,24 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
+
+	private function isSupportedByWater() : bool{
+		$down = $this->getSide(Facing::DOWN);
+		foreach(Facing::HORIZONTAL as $side){
```
https://github.com/pmmp/PocketMine-MP/blob/b3b8aaddffe3b3a1eee8dcc62dd08fcd10581916/src/block/Block.php#L788

bonbionseker replied:
okay let's try this

Commit:
Code Style Fix #2

Commit:
Update Sugarcane.php

bonbionseker reviewed (commented):
._.

Commit:
Revert "Code Style Fix #2"

This reverts commit b75e6b8855c5bfa7ac045e868536020acf9a53eb.

Commit:
Update Sugarcane.php

Commit:
Update Sugarcane.php

Commit:
FixCodeStyle

pls fix :C

Commit:
FixCodeStyle

Commit:
onNearbyBlockChange

No, only the sugar cane underneath will be checked.

Commit:
Code Style Fix

:C

Commit:
Merge branch 'stable' into bonbionseker-patch-1

ipad54 commented on src/block/Sugarcane.php:
```diff
@@ -98,6 +98,12 @@ public function ticksRandomly() : bool{
 		if(!$this->getSide(Facing::DOWN)->hasSameTypeId($this)){
+			if(!$this->isSupportedByWater()) {
+				$world = $this->position->getWorld();
+				$world->useBreakOn($this->position, createParticles: true);
```
There is no need to create an additional `$world` variable. You can inline this.

ipad54 commented on src/block/Sugarcane.php:
```diff
@@ -123,4 +129,28 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
+		$down = $this->getSide(Facing::DOWN);
+		foreach($down->getHorizontalSides() as $sideBlock){
+			$blockId = $sideBlock->getTypeId();
+
```
```suggestion
```

ipad54 commented on src/block/Sugarcane.php:
```diff
@@ -123,4 +129,28 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
+
+	public function onNearbyBlockChange() : void {
+		$bottom = $this->seekToBottom();
+		if (!$this->position->equals($bottom)) {
```
It would be better if you combine these 2 conditions into one (e.g `if($this->position->equals($bottom) && !$this->isSupportedByWater()){}`)

ipad54 replied:
By the way, you can just check that the down side of the block doesn't have same type id instead of seeking to the bottom (like it is done in `onRandomTick()`). It would be a little bit faster in large crop farms.

ipad54 commented on src/block/Sugarcane.php:
```diff
@@ -123,4 +129,28 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
+		}
+
+		if (!$this->isSupportedByWater()) {
+			$world = $this->position->getWorld();
```
Same here. There is no need to create additional `$world` variable.

Commit:
Adjust Sugar Cane to break when there is no water

I edited the events that needed editing :) thanks ipad54

dktapps commented on src/block/Sugarcane.php:
```diff
@@ -123,4 +128,22 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
 		return false;
 	}
+
+	private function isSupportedByWater() : bool{
```
`hasNearbyWater` would be a better name for this

dktapps commented on src/block/Sugarcane.php:
```diff
@@ -123,4 +128,22 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
 	}
+
+	private function isSupportedByWater() : bool{
+		$down = $this->getSide(Facing::DOWN);
```
`getSide(down)` is getting called outside of where this is used in both usages. Perhaps it might be worth passing it as a parameter instead of calling it a second time.

dktapps commented on src/block/Sugarcane.php:
```diff
@@ -123,4 +128,22 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
+		$down = $this->getSide(Facing::DOWN);
+		foreach($down->getHorizontalSides() as $sideBlock){
+			$blockId = $sideBlock->getTypeId();
+			if ($blockId === BlockTypeIds::WATER || $blockId === BlockTypeIds::FROSTED_ICE) {
```
```suggestion
			if($blockId === BlockTypeIds::WATER || $blockId === BlockTypeIds::FROSTED_ICE){
```

dktapps commented on src/block/Sugarcane.php:
```diff
@@ -123,4 +128,22 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
+
+	public function onNearbyBlockChange() : void {
+		$downBlock = $this->getSide(Facing::DOWN);
+		if (!$downBlock->hasSameTypeId($this) && !$this->isSupportedByWater()) {
```
```suggestion
		if(!$downBlock->hasSameTypeId($this) && !$this->isSupportedByWater()){
```

dktapps commented on src/block/Sugarcane.php:
```diff
@@ -98,6 +98,11 @@ public function ticksRandomly() : bool{
 
 	public function onRandomTick() : void{
 		if(!$this->getSide(Facing::DOWN)->hasSameTypeId($this)){
+			if(!$this->isSupportedByWater()) {
```
```suggestion
			if(!$this->isSupportedByWater()){
```

Commit:
Adjust Sugar Cane to break when there is no water

Ok I fixed these too :)

Commit:
Adjust Sugar Cane to break when there is no water

I forgot these ._.

Commit:
Adjust Sugar Cane to break when there is no water

Okay done! 👍

Commit:
Merge branch 'minor-next' into bonbionseker-patch-1

kostamax27 reviewed (commented):
looks good

ipad54 commented on src/block/Sugarcane.php:
```diff
@@ -123,4 +129,21 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
+		return false;
+	}
+
+	public function onNearbyBlockChange() : void {
```
Sugarcane uses `StaticSupportTrait`, so overriding `onNearbyBlockChange()` without calling trait`s method will make sugarcane to stay after supporting block removed.

![Screenshot_20241113_111810](https://github.com/user-attachments/assets/d61fac7c-8f2b-49b9-a8a7-2544d63fd54f)

bonbionseker replied:
I don't know how to find a solution for this  :C

ipad54 replied:
You can import trait's method as alias and call it after your code. Something like [here](https://github.com/pmmp/PocketMine-MP/blob/09bf203267aca9b83a09640e17bfac0fc7b58ff8/src/block/PinkPetals.php#L40) but with `onNearbyBlockChange()`

dktapps replied:
Good catch @ipad54, I didn't spot that

ipad54 commented on src/block/Sugarcane.php:
```diff
@@ -123,4 +129,21 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
+		return false;
+	}
+
+	public function onNearbyBlockChange() : void {
```
```suggestion
	public function onNearbyBlockChange() : void{
```

Also formatting issue.

Commit:
Adjust Sugar Cane to break when there is no water

formatting issue fix! 👍

Commit:
Update Sugarcane.php

By adding "canBeSupportedAt" I made it break when we broke the support block. -.-

Commit:
Update Sugarcane.php

formatting issue fix! 👍

Commit:
I removed "canBeSupportedAt" and tried to use "StaticSupportTrait" instead

Okay, I found how to use it and implemented it. 👍

Commit:
Code Style Fix

Did you come again, no it can't be :C

Commit:
Merge branch 'minor-next' into bonbionseker-patch-1

ipad54 commented on src/block/Sugarcane.php:
```diff
@@ -36,7 +36,9 @@
 	use AgeableTrait;
-	use StaticSupportTrait;
+	use StaticSupportTrait{
+		onNearbyBlockChange as onNearbyBlockChangeTrait;
```
Better to name it `onSupportBlockChange`

ipad54 commented on src/block/Sugarcane.php:
```diff
@@ -36,7 +36,9 @@
 class Sugarcane extends Flowable{
 	use AgeableTrait;
-	use StaticSupportTrait;
+	use StaticSupportTrait{
```
```suggestion
	use StaticSupportTrait {
```

ipad54 commented on src/block/Sugarcane.php:
```diff
@@ -123,4 +131,22 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
+		$down = $this->getSide(Facing::DOWN);
+		if(!$down->hasSameTypeId($this) && !$this->hasNearbyWater($down)){
+			$this->position->getWorld()->useBreakOn($this->position, createParticles: true);
+		}
```
Please add `else` here - we need to make sure the block won't be destroyed twice.

bonbionseker replied:
```
if(!$down->hasSameTypeId($this) && !$this->hasNearbyWater($down)) {
            $this->position->getWorld()->useBreakOn($this->position, createParticles: true);
        } else {
			//Won't be destroyed twice.
	}
```
Is this exactly how it should be? 🤕

ipad54 replied:
```php
 if(!$down->hasSameTypeId($this) && !$this->hasNearbyWater($down)){
            $this->position->getWorld()->useBreakOn($this->position, createParticles: true);
        }else{
			$this->onSupportBlockChange();
	}
}```

Commit:
Update Sugarcane.php

Name change and formatting issue fix!

Commit:
Update Sugarcane.php

else

Commit:
Merge branch 'minor-next' into bonbionseker-patch-1
